File: backend/courts/application/use_cases/get_weekly_availability.py
# backend/courts/application/use_cases/get_weekly_availability.py

from datetime import datetime, timedelta

class GetWeeklyAvailabilityUseCase:
    """
    Caso de uso para obtener la disponibilidad semanal de una cancha específica.
    """
    def __init__(self, court_repository):
        """
        Inicializa el caso de uso con un repositorio de canchas.

        Args:
            court_repository: El repositorio de canchas que implementa la interfaz CourtRepository.
        """
        self.court_repository = court_repository

    async def execute(self, court_id: int, start_date: datetime, end_date: datetime):
        """
        Ejecuta el caso de uso para obtener la disponibilidad semanal.

        Args:
            court_id (int): El ID de la cancha.
            start_date (datetime): La fecha y hora de inicio del rango semanal.
            end_date (datetime): La fecha y hora de fin del rango semanal.

        Returns:
            dict: Un diccionario que representa la disponibilidad semanal de la cancha.
                  Ejemplo: { 'YYYY-MM-DD': { hour: boolean, ... }, ... }
        """
        # La lógica para obtener la disponibilidad detallada por hora y día
        # se delegará al repositorio.
        try:
            weekly_availability = await self.court_repository.get_weekly_availability(
                court_id, start_date, end_date
            )
            return weekly_availability
        except Exception as e:
            print(f"Error en GetWeeklyAvailabilityUseCase: {e}")
            raise

File: frontend/src/application/use-cases/courts/get-weekly-availability.js
import { ICourtRepository } from '../../../domain/repositories/court-repository'; // Importar la interfaz del repositorio

/**
 * Caso de uso para obtener la disponibilidad semanal de una cancha.
 * Esta clase reside en la capa de Aplicación y orquesta la obtención de datos
 * utilizando la interfaz del repositorio de Dominio.
 */
export class GetWeeklyAvailabilityUseCase {
  /**
   * @param {ICourtRepository} courtRepository - Una implementación del repositorio de canchas.
   */
  constructor(courtRepository) {
    if (!(courtRepository instanceof ICourtRepository)) {
      throw new Error('courtRepository must be an instance of ICourtRepository');
    }
    this.courtRepository = courtRepository;
  }

  /**
   * Ejecuta el caso de uso para obtener la disponibilidad semanal.
   * @param {number} courtId - El ID de la cancha.
   * @param {string} start_date - La fecha de inicio de la semana (ISO 8601).
   * @param {string} end_date - La fecha de fin de la semana (ISO 8601).
   * @returns {Promise<object>} Una promesa que resuelve con un objeto que representa la disponibilidad semanal.
   */
  async execute(courtId, start_date, end_date) {
    // Aquí se podría añadir lógica de aplicación adicional si fuera necesario.
    return this.courtRepository.getWeeklyAvailability(courtId, start_date, end_date);
  }
}
